### PR TITLE
py-dask-ml: add new package.

### PR DESCRIPTION
**Pull request by t-karatsu**
Add new package `dask-ml`.

- Described dependencies of `+docs` are based on the following files:
https://github.com/dask/dask-ml/blob/main/ci/environment-docs.yaml
- `xgboost_dependency.patch` is for modifying `xgboost`'s dependencies according to the following issues:
https://github.com/dask/dask-ml/issues/797
- When using `aarch64` environment and `gcc` compiler, a warning about `numpy` occurs. When installing `py-dask-ml` with `+docs`, the warning is treated as an error and cannot be installed. So, I added conflict.

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-dask-ml/package.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class PyDaskMl(Package):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-dask-ml/package.py]:
```suggestion
class PyDaskMl(PythonPackage):
```

File: var/spack/repos/builtin/packages/py-dask-ml/package.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    """Scalable Machine Learning with Dask."""
+
+    homepage = "https://ml.dask.org/"
+    url      = "https://pypi.io/packages/source/d/dask-ml/dask-ml-1.8.0.tar.gz"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-dask-ml/package.py]:
```suggestion
    pypi     = "dask-ml/dask-ml-1.8.0.tar.gz"
```

File: var/spack/repos/builtin/packages/py-dask-ml/package.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+    version('1.8.0', sha256='8fc4ac3ec1915e382fb8cae9ff1ec9b5ac1bee0b6f4c6975d6e6cb7191a4a815')
+
+    variant('docs', default=False, description='Build HTML documentation')
+    variant('xgboost', default=False, description='Deploys XGBoost alongside Dask')
+
+    extends('python')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-dask-ml/package.py]: Once you subclass PythonPackage this line can be removed

File: var/spack/repos/builtin/packages/py-dask-ml/package.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    depends_on('python@3.6:', type=('build', 'run'))
+
+    depends_on('py-setuptools', type='build')
+    depends_on('py-setuptools-scm', type='build')
+    depends_on('py-pip', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-dask-ml/package.py]: Not needed if you use the normal PythonPackage build system

File: var/spack/repos/builtin/packages/py-dask-ml/package.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    depends_on('py-sparse', type=('build', 'run'), when='+docs')
+    depends_on('py-sphinx', type=('build', 'run'), when='+docs')
+    depends_on('py-sphinx-rtd-theme', type=('build', 'run'), when='+docs')
+    depends_on('py-sphinx-gallery', type=('build', 'run'), when='+docs')
+    depends_on('py-testpath', type=('build', 'run'), when='+docs')
+    depends_on('py-tornado', type=('build', 'run'), when='+docs')
+    depends_on('py-toolz', type=('build', 'run'), when='+docs')
+    depends_on('py-zict', type=('build', 'run'), when='+docs')
+    depends_on('py-dask-sphinx-theme@1.1.0:', type=('build', 'run'), when='+docs')
+    depends_on('py-nbsphinx', type=('build', 'run'), when='+docs')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-dask-ml/package.py]: Many of these come from `test_requires`, not `doc_requires`. I think this list can be shortened.

File: var/spack/repos/builtin/packages/py-dask-ml/package.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+    depends_on('py-xgboost+dask', type=('build', 'run'), when='+docs')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-dask-ml/package.py:53]: Why is this needed to build the docs?

File: var/spack/repos/builtin/packages/py-dask-ml/package.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    conflicts('+docs', when='%gcc target=aarch64:')
+
+    def install(self, spec, prefix):
+        pip = which('pip')
+
+        if '+xgboost' in spec:
+            pip('install', self.stage.source_path, '.[xgboost]',
+                '--prefix={0}'.format(prefix))
+        else:
+            pip('install', self.stage.source_path, '--prefix={0}'.format(prefix))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-dask-ml/package.py]: The default build/install phases of PythonPackage should work

File: var/spack/repos/builtin/packages/py-dask-ml/package.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+        if '+docs' in spec:
+            with working_dir('docs'):
+                make('html')
+                install_tree(
+                    join_path(self.stage.source_path, 'docs'), self.prefix.docs)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-dask-ml/package.py]: I would do something like:
```python
@run_after('build')
def build_docs(self):
    if '+docs' in self.spec:
        with working_dir('docs'):
            make('html')

@run_after('install')
def install_docs(self):
    if '+docs' in self.spec:
        install_tree('docs', self.prefix.docs)
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-dask-ml/package.py (hunk 1 above)
Comment (t-karatsu) [reply var/spack/repos/builtin/packages/py-dask-ml/package.py]: I modified it to `PythonPackage` format.

File: var/spack/repos/builtin/packages/py-dask-ml/package.py (hunk 2 above)
Comment (t-karatsu) [reply var/spack/repos/builtin/packages/py-dask-ml/package.py]: I modified it to `PythonPackage` format.

File: var/spack/repos/builtin/packages/py-dask-ml/package.py (hunk 3 above)
Comment (t-karatsu) [reply var/spack/repos/builtin/packages/py-dask-ml/package.py]: I removed it.

File: var/spack/repos/builtin/packages/py-dask-ml/package.py (hunk 4 above)
Comment (t-karatsu) [reply var/spack/repos/builtin/packages/py-dask-ml/package.py]: I removed it.

File: var/spack/repos/builtin/packages/py-dask-ml/package.py (hunk 7 above)
Comment (t-karatsu) [reply var/spack/repos/builtin/packages/py-dask-ml/package.py]: I removed it.

File: var/spack/repos/builtin/packages/py-dask-ml/package.py (hunk 8 above)
Comment (t-karatsu) [reply var/spack/repos/builtin/packages/py-dask-ml/package.py]: I changed to suggested functions.

File: var/spack/repos/builtin/packages/py-dask-ml/package.py (hunk 6 above)
Comment (t-karatsu) [reply var/spack/repos/builtin/packages/py-dask-ml/package.py:53]: This is needed to output `xgboost` related files. If not specified, some files will not be output.

File: var/spack/repos/builtin/packages/py-dask-ml/package.py (hunk 8 above)
Comment (t-karatsu) [reply var/spack/repos/builtin/packages/py-dask-ml/package.py]: I'm having problems not finding the `dask-ml` module similar to the following issue, so I'll change the docs build after install.
https://github.com/dask/dask-ml/issues/693

File: var/spack/repos/builtin/packages/py-dask-ml/package.py (hunk 5 above)
Comment (t-karatsu) [reply var/spack/repos/builtin/packages/py-dask-ml/package.py]: I removed test-related dependencies.

**Comment by t-karatsu** [comment]
@adamjstewart 
Thank you for your review. I have corrected the pointed out part, could you confirm it again?

**Review by adamjstewart — APPROVED** [review]
